controller/backing_image_controller.go

```diff
@@ -310,22 +345,530 @@ func (bic *BackingImageController) syncBackingImage(key string) (err error) {
+	// For case 1 and case 2, the status will come back automatically when server pick the snapshot lvol up and reconstruct the record.
+	// We delete the unknown copy only for the case 3 if there is a ready copy so it can be recreated.
+	if !hasReadyV2BackingImage {
+		return nil
+	}
+
+	for _, v2DiskUUID := range v2CopiesUnknown {
+		node, _, err := bic.ds.GetReadyDiskNodeRO(v2DiskUUID)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get the ready disk node for disk %v", v2DiskUUID)
+		}
+
+		isDeleted, err := bic.deleteUnknownV2Copy(bi, v2DiskUUID, node, log)
+		if err != nil {
+			return errors.Wrapf(err, "failed to delete the unknown v2 copy on disk %v", v2DiskUUID)
+		}
+		// lvol is found so we didn't delete the unknown copy
+		if !isDeleted {
+			continue
+		}
+
+		delete(bi.Status.DiskFileStatusMap, v2DiskUUID)
+		bic.v2CopyBackoff.DeleteEntry(v2DiskUUID)
+		bic.eventRecorder.Eventf(bi, corev1.EventTypeNormal, constant.EventReasonDelete, "Deleted unknown v2 backing image %v on disk %v on node %v", bi.Name, v2DiskUUID, node.Name)
+	}
+
+	for _, v2DiskUUID := range v2CopiesFailed {
+		node, _, err := bic.ds.GetReadyDiskNodeRO(v2DiskUUID)
+		if err != nil {
+			return errors.Wrapf(err, "failed to get the ready disk node for disk %v", v2DiskUUID)
+		}
+
+		log.Infof("Deleting the failed v2 copy on disk %v", v2DiskUUID)
+		if err := bic.deleteV2Copy(bi, v2DiskUUID, node, log); err != nil {
+			return errors.Wrapf(err, "failed to delete the failed v2 copy on disk %v", v2DiskUUID)
+		}
+
+		delete(bi.Status.DiskFileStatusMap, v2DiskUUID)
+		bic.v2CopyBackoff.DeleteEntry(v2DiskUUID)
+		bic.eventRecorder.Eventf(bi, corev1.EventTypeNormal, constant.EventReasonDelete, "Deleted failed v2 backing image %v on disk %v on node %v", bi.Name, v2DiskUUID, node.Name)
+	}
+	return nil
+}
+
+func (bic *BackingImageController) cleanupAllV2BackingImageCopies(bi *longhorn.BackingImage) (cleaned bool, err error) {
+	log := getLoggerForBackingImage(bic.logger, bi)
+
+	if err := bic.syncV2StatusWithInstanceManager(bi); err != nil {
+		return false, errors.Wrapf(err, "failed to sync v2 backing image status from instance manager")
+	}
+
+	// we can delete the backing image when all the v2 copies are deleted
+	if len(bi.Status.DiskFileStatusMap) == 0 {
+		return true, nil
+	}
+
+	allV2CopiesCleaned := true
+	for diskUUID := range bi.Status.DiskFileStatusMap {
+		if fileSpec, exists := bi.Spec.DiskFileSpecMap[diskUUID]; exists && !types.IsDataEngineV2(fileSpec.DataEngine) {
+			continue
+		}
+
+		allV2CopiesCleaned = false
+
+		node, _, err := bic.ds.GetReadyDiskNodeRO(diskUUID)
+		if err != nil {
+			return false, errors.Wrapf(err, "failed to get the ready disk node for disk %v", diskUUID)
+		}
+
+		log.Infof("Deleting the v2 copy on disk %v for backing image %v", diskUUID, bi.Name)
+		if err := bic.deleteV2Copy(bi, diskUUID, node, log); err != nil {
+			return false, errors.Wrapf(err, "failed to delete the v2 copy on disk %v", diskUUID)
+		}
```

Comment:
Does it handle (ignore) `not found` error?

Reply:
Yes, in longhorn-spdk-server